source/util/FeatureRegistry.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "FeatureRegistry.h"
+#include "../logging/LoggerFactory.h"
+
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    for (auto &feature : features)
+    {
+        if (feature->getName() == name)
+        {
+            return feature;
+        }
+    }
+
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::shared_ptr<Feature> feature)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (std::find(features.begin(), features.end(), feature) != features.end())
```

Comment:
This loop will search for features using `operator==`. In this case that means feature is considered equal when address are equal. I am not sure whether that is what you intended and the behavior of `add` is not symmetric with `get` since the latter matches by name. If you instead wanted to search for feature matching by name, then you need something like this.

```
auto result = std::find_if(
    std::begin(features), 
    std::end(features), 
    [&feature](auto f) { return f->getName() == feature->getName(); }
);
```